### PR TITLE
Efficient scan of jandex index for Spring JPA Annotations

### LINKED ISSUE #19323: Degredation in Dev mode startup time for project using quarkus-spring-jpa extension
### Describe the bug

Time to first request has degraded for dev mode startup between 1.11.7.Final and 2.1.1.Final for projects that include a lot of extensions (i.e. many classes) that also include `quarkus-spring-data-jpa`

Start times reported by the console;

1.11.7.Final | 2.1.1.Final
-- | --
3.954s. | 5.804s.
3.923s. | 6.397s.
3.949s. | 6.408s.
4.054s. | 5.780s.
4.111s. | 5.980s.

This also impacts time to first response in dev mode. The issue is caused by an inefficient scan of Jandex indexed classes while searching for classes annotated with `org.springframework.data.repository.NoRepositoryBean`

Between 1.11.7.Final and 2.1.1.Final, this scan has become more costly and therefore start times have regressed.

[profiles-cpu.zip](https://github.com/quarkusio/quarkus/files/6963547/profiles-cpu.zip)

### Expected behavior

_No response_

### Actual behavior

Dev mode start time degrades are more classes are added to the classpath

### How to Reproduce?

**2.1.1.Final**

generate app by running the following command:

`mvn io.quarkus:quarkus-maven-plugin:2.1.1.Final:create -DprojectGroupId=my-groupId -DprojectArtifactId=app-generated-skeleton -DprojectVersion=1.0.0-SNAPSHOT -DpackageName=org.my.group -DplatformArtifactId=quarkus-bom -DplatformVersion=2.1.1.Final -Dextensions=agroal,config-yaml,core,hibernate-orm,hibernate-orm-panache,hibernate-validator,jackson,jaxb,jdbc-mysql,jdbc-postgresql,jsonb,jsonp,kafka-client,logging-json,narayana-jta,oidc,quarkus-quartz,reactive-pg-client,rest-client,rest-client-jaxb,resteasy,resteasy-jackson,resteasy-jaxb,resteasy-jsonb,scheduler,spring-boot-properties,smallrye-reactive-streams-operators,spring-data-jpa,spring-di,spring-security,spring-web,undertow,vertx,vertx-web,grpc,cache,micrometer,quarkus-openshift-client`

And then

`mvn clean quarkus:dev -Dquarkus.kafka.devservices.enabled=false -Dquarkus.keycloak.devservices.enabled=false`

**1.11.7.Final**

`mvn io.quarkus:quarkus-maven-plugin:1.11.7.Final:create -DprojectGroupId=my-groupId -DprojectArtifactId=app-generated-skeleton -DprojectVersion=1.0.0-SNAPSHOT -DpackageName=org.my.group -DplatformArtifactId=quarkus-bom -DplatformVersion=1.11.7.Final -Dextensions=agroal,config-yaml,core,hibernate-orm,hibernate-orm-panache,hibernate-validator,jackson,jaxb,jdbc-mysql,jdbc-postgresql,jsonb,jsonp,kafka-client,logging-json,narayana-jta,quarkus-quartz,reactive-pg-client,rest-client,rest-client-jaxb,resteasy,resteasy-jackson,resteasy-jaxb,resteasy-jsonb,scheduler,spring-boot-properties,smallrye-reactive-streams-operators,spring-data-jpa,spring-di,spring-security,spring-web,undertow,vertx,vertx-web,grpc,cache,micrometer,quarkus-openshift-client`

And then:

`mvn clean quarkus:dev -Dquarkus.log.console.json=false`

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by johnaohara**
Fixes #19323

### CODE REVIEW

**Comment by johnaohara** [comment]
Startup cpu profile with this fix;

[profile-999-SNAPSHOT.zip](https://github.com/quarkusio/quarkus/files/6963615/profile-999-SNAPSHOT.zip)

File: extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/SpringDataJPAProcessor.java [hunk 1]
```diff
@@ -232,20 +233,12 @@ private void removeNoRepositoryBeanClasses(List<ClassInfo> interfacesExtendingRe
...
-                if (classInfo.classAnnotation(DotNames.SPRING_DATA_NO_REPOSITORY_BEAN) != null) {
-                    result.add(classInfo.name());
-                }
-            }
-        }
-        return result;
+        return index.getKnownClasses()
+                .stream()
+                .filter(classInfo -> Modifier.isInterface(classInfo.flags()))
+                .filter(classInfo -> classInfo.classAnnotation(DotNames.SPRING_DATA_NO_REPOSITORY_BEAN) != null)
```
Comment (gsmet) [inline extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/SpringDataJPAProcessor.java:239]: I might be missing something but why don't we search for the annotation instead? And then filter the target to be a class and then check the interfaces?

**Review by gsmet — CHANGES_REQUESTED** [review]

File: extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/SpringDataJPAProcessor.java (hunk 1 above)
Comment (gsmet) [reply extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/SpringDataJPAProcessor.java]: @johnaohara I pushed another commit with an implementation using the index. Can you check it's good for you? I think it should be better as we don't end up scanning the whole index.

Comment (johnaohara) [reply extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/SpringDataJPAProcessor.java]: @gsmet I was just testing a change similar to what you pushed. It looks good, but I'll pull down your changes and verify against the start up test

Comment (johnaohara) [reply extensions/spring-data-jpa/deployment/src/main/java/io/quarkus/spring/data/deployment/SpringDataJPAProcessor.java]: It is good,  CPU time in this method has gone from 24.7% in 2.1.1.Final to 0.4% with this PR

**Comment by johnaohara** [comment]
Updated Profile with https://github.com/quarkusio/quarkus/commit/1b56fa839ecfcfc6fac98e8691eb65965f945e84
[profile-1b56fa839ecfcfc6fac98e8691eb65965f945e84.zip](https://github.com/quarkusio/quarkus/files/6964313/profile-1b56fa839ecfcfc6fac98e8691eb65965f945e84.zip)

**Review by gsmet — APPROVED** [review]

**Pull request merged**